src/App.js

```diff
@@ -24,62 +24,87 @@ import getToken from './lib/tokens';
     const token = getToken();
-    localStorage.setItem('token', token);
-    setUserToken(token);
+
+    return db
+      .collection('lists')
+      .add({ token: token })
+      .then((results) => {
+        const { id } = results;
+        if (id === undefined) throw new Error('Failed to create list.');
+        return id;
+      })
+      .then((newListId) => {
+        // list successfully created in db, so update listId state and save token to localStorage
+        setListId(newListId);
+        localStorage.setItem('token', token);
+        return true;
+      });
+  }
+
+  // returns true if token matches a list in db, returns false if no list, throws error on connection problem
+  function isTokenValid(token) {
+    return db
+      .collection('lists')
+      .where('token', '==', token)
+      .get()
+      .then((querySnapshot) => {
+        // if there are results and an id property exists
+        if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
+          setListId(querySnapshot.docs[0].id);
+          return true;
+
+          // check metadata.fromCache to distinguish between no results (invalid token) and a connection issue
+        } else if (!querySnapshot.metadata.fromCache) {
+          return false;
+        } else {
+          throw new Error('Connection problem');
+        }
+      });
+  }
+
+  function joinList(shareToken) {
+    return isTokenValid(shareToken).then((listExists) => {
+      if (listExists) {
+        localStorage.setItem('token', shareToken);
+        return true;
+      } else throw new Error('Invalid token');
+    });
   }
 
   // on component mounting, look for token in local storage and use it to retrieve the list id
   useEffect(() => {
     const token = localStorage.getItem('token');
 
     if (token) {
-      setUserToken(token);
-
-      // find the list in Firestore associated with the stored token from local storage
-      db.collection('lists')
-        .where('token', '==', token)
-        .get()
-        .then((querySnapshot) => {
-          // if there are results and an id property exists
-          if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
-            setListId(querySnapshot.docs[0].id); // save the list id for later
-          }
+      isTokenValid(token)
+        .then((listExists) => {
+          if (listExists !== true) localStorage.removeItem('token');
         })
         .catch((error) => {
           console.log('Error getting list: ', error);
         });
     }
-  }, [listId]);
+  }, []);
 
   return (
     <Router>
       <div className="App container">
         <Switch>
           <Route exact path="/">
-            {userToken ? (
+            {listId ? (
               <Redirect to="/list" />
             ) : (
-              <Home
-                userToken={userToken}
-                setUserToken={setUserToken}
-                saveToken={saveToken}
-              />
+              <Home createList={createList} joinList={joinList} />
```

Comment:
Is there a reason why we can't have the `createList` and `joinList` functions in Home.js instead of passing them down as props in App.js? 

There may be an architectural or best practice reason in React, but it was hard for me to jump between Home.js and App.js to understand what the functions were doing.

Reply:
Yes, good observation! This is mostly to do with the way useHistory works, and partly because it made sense to maintain error states and values for the Home view forms inside the Home component. 

It may help to think of it like this: 99% of the work for creating a list or joining a list is handled by the functions declared in App.js. Inside the Home component, we want to use these functions when our "Create List" button is pushed, or when our "Join List" form is submitted, but we also need to do a couple of "local" things in Home based on the results: 
1. We need to handle redirects in Home, because useHistory won't work properly in App.js (I believe @nick-zanetti has a link with more info on this behaviour⁠—Nick, can you share?)
2. We need to update the error state for the Join List form to show error messages to the user, and seemed to make the most sense to keep this state in the Home component.

We added `handleCreateList` and `handleJoinList` functions to handle events in Home, and they simply run the `createList`/`joinList` functions then do the extra bits above.

Reply:
So to expand on what Ander was saying, a problem Sheila and I ran into last week was that `useHistory()` must be called in a _child functional component_. The [history instance](https://reactrouter.com/web/api/history) (which `useHistory` gives us access to) uses a context that is set by the `<Router>` component (in our case, I think this context pretty much means the URL slugs/routes). If we call `useHistory()` in App.js, that context will not be set in time for us to use it, so it will be undefined. More on this [here](https://stackoverflow.com/questions/62614433/react-router-why-is-the-usehistory-undefined-in-react). I'm still trying to wrap my head around this so hopefully I didn't butcher that explanation 😅

Reply:
Yes I remember this issue that Nick and I had. After some research, I believe we found where it would only work in child components, and that it won't work when it's within the same components where the Routes are listed.

Reply:
Actually @nick-zanetti, I think the way you put it makes perfect sense! Very good explanation.